crates/sui-framework/sources/crypto/drand.move

```diff
@@ -0,0 +1,220 @@
+/// WARNING: Using this module places the drand committee (and all Drand committees of the past) on
+/// a) the shared secret's secrecy (no collusion)
+/// b) producing new rounds every PERIOD seconds (only relevant for using `drand::unsafe_unix_time_seconds` as a time reference
+module sui::drand {
+    use std::hash::sha2_256;
+    use std::vector;
+
+    use sui::bls12381;
+    use sui::object::{Self, UID};
+    use sui::transfer;
+    use sui::tx_context::TxContext;
+
+    // Error codes
+    /// Expected a 16 byte RND, but got a different length
+    const EInvalidRndLength: u64 = 0;
+    /// Could not verify signature on Drand
+    const EInvalidProof: u64 = 1;
+    /// Trying to advance to a round that has already passed
+    const ERoundAlreadyPassed: u64 = 2;
+
+    // Constants
+    const CHAIN: vector<u8> = x"8990e7a9aaed2ffed73dbd7092123d6f289930540d7651336225dc172e51b2ce";
+    /// The genesis time of CHAIN
+    const GENESIS: u64 = 1595431050;
+    /// The public key of CHAIN
+    const DRAND_PK: vector<u8> =
+        x"868f005eb8e6e4ca0a47c8a77ceaa5309a47978a7c71bc5cce96366b5d7a569937c529eeda66c7293784a9402801af31";
+    /// Time in seconds between randomness beacon rounds for CHAIN
+    const PERIOD: u64 = 30;
+
+    /// Shared object recording drand progress. Anyone can advance the round by providing a valid signature
+    struct Drand has key {
+        id: UID,
+        /// Most recent drand round we have seen
+        previous_round: u64,
+        /// Signature from the last drand round
+        previous_sig: vector<u8>,
+    }
+
+    fun init(ctx: &mut TxContext) {
+        // initialize at a fairly recent round
+        transfer::share_object(
+            Drand {
+                id: object::new(ctx),
+                previous_round: 2475343,
+                previous_sig: x"880c20aa83669a234f7b633917e43b7afb4e72421f0f520f7f570559356ed72e15867a77513346fbd37401709fc2192c1830936f23a751c228d091527290b304f2c6fa93851ca92c5ae7c3f7d7e0d06c65b561a6d6afeabce080f442f68d3845"
+            }
+        )
+    }
+
+    /// Advance `drand` by one round
+    public fun advance(drand: &mut Drand, sig: vector<u8>) {
+        let round = drand.previous_round + 1;
+        verify_drand_signature(sig, drand.previous_sig, round);
+        drand.previous_sig = sig;
+        drand.previous_round = round;
+    }
+
+    /// Advance `drand` by an arbitrary number of rounds.
+    /// Aborts if `round` is smaler than the most recent round `drand` has seen
+    public fun jump(drand: &mut Drand, sig: vector<u8>, previous_sig: vector<u8>, round: u64) {
+        assert!(round > drand.previous_round, ERoundAlreadyPassed);
+        verify_drand_signature(sig, previous_sig, round);
+        drand.previous_sig = sig;
+        drand.previous_round = round;
+    }
+
+    /// Return the most recent round we have seen
+    public fun round(drand: &Drand): u64 {
+        drand.previous_round
```

Comment:
a comment maybe for @benr-ml that naming it `previous_round` might be confusing, especially when after we register it, someone (before the next round being published) tries to get the current round seed. In theory it's "most recent registered round". We'll fix wording here in an upcoming PR.